source/slang/slang-lower-to-ir.cpp

```diff
@@ -7859,21 +7863,30 @@ struct SpecializedComponentTypeIRGenContext : ComponentTypeVisitor
 
                 IRInst* irParam = getSimpleVal(context, ensureDecl(context, shaderParam.paramDeclRef));
                 List<IRInst*> irSlotArgs;
+                // Tracks if there are any type args that is not an IRDynamicType.
+                bool hasConcreteTypeArg = false;
                 for( Index jj = 0; jj < specializationArgCount; ++jj )
                 {
                     auto& specializationArg = specializationInfo->existentialArgs[existentialArgOffset++];
 
                     auto irType = lowerSimpleVal(context, specializationArg.val);
                     auto irWitness = lowerSimpleVal(context, specializationArg.witness);
 
+                    if (irType->op != kIROp_DynamicType)
+                        hasConcreteTypeArg = true;
+
                     irSlotArgs.add(irType);
                     irSlotArgs.add(irWitness);
                 }
-
-                builder->addBindExistentialSlotsDecoration(
-                    irParam,
-                    irSlotArgs.getCount(),
-                    irSlotArgs.getBuffer());
+                // Only insert a `BindExistentialSlots` decoration when there are at least
+                // one non-dynamic type argument.
+                if (hasConcreteTypeArg)
+                {
+                    builder->addBindExistentialSlotsDecoration(
+                        irParam,
+                        irSlotArgs.getCount(),
+                        irSlotArgs.getBuffer());
+                }
```

Comment:
The existential specialization pass should incrementally make the granularity of `BindExistentials<...>` smaller and smaller, until eventually everything is an `ExistentialBox<...>`.

It should be possible to add logic to the existential specialization pass that notices when we are about to introduce an `ExistentialBox<__Dynamic, IFoo>` and decides *not* to introduce that box and directly use `IFoo` instead. That would seemingly help in the case where a top-level type is being specialized for some parameters and not others.

Reply:
Agreed, I was trying to avoid touching too many fronts in this PR. This logic here doesn't hurt even when we make the bind-existential pass capable of ha doing partial specialization.